src/runtime_src/core/common/drv/kds_core.c

```diff
@@ -68,13 +103,37 @@ ssize_t show_kds_custat_raw(struct kds_sched *kds, char *buf)
 			if (xcu->info.slot_idx != j)
 				continue;
 
-			sz += scnprintf(buf+sz, PAGE_SIZE - sz, cu_fmt, j,
+			/* Generate the CU string to write into the buffer */
+			memset(cu_buf, 0, sizeof(cu_buf));
+			cu_sz = kds_create_cu_string(xcu, &cu_buf, j,
 					set_domain(DOMAIN_PL, i),
-					xcu->info.kname, xcu->info.iname,
-					xcu->info.addr, xcu->status,
-					cu_stat_read(cu_mgmt, usage[i]));
+					cu_stat_read(cu_mgmt, usage[i]), type);
+
+			/* Store the CU string length with previous lengths */
+			all_cu_sz += cu_sz;
+
+			/**
+			 * Verify that
+			 * 1. The buffer can hold the data
+			 * 2. The data starts after the requested offset
+			 */
+			if (sz + cu_sz > buf_size)
+				return sz;
+
+			if (all_cu_sz > offset)
+				sz += scnprintf(buf+sz, buf_size - sz, "%s", cu_buf);
```

Comment:
don't you need to first check if the data needs to be transferred before check if the buffer is bug enough to hold it? Basically check condition 2, then 1?